### PR TITLE
Relax the constraints of isValidBaseType to allow base types to be constructor types

### LINKED ISSUE #33127: A class can't extend from object/generic intersection, but can extend from plain generic
Ran into this looking into an issue with @athasach:

```ts
export type Constructor<T extends object = object> = new (...args: any[]) => T;

export interface Initable {
    init(...args: any[]): void;
}

/**
 * Plain mixin where the superclass must be Initable
 */
export const Serializable = <K extends Constructor<Initable> & Initable>(
    SuperClass: K
) => {
    const LocalMixin = (InnerSuperClass: K) => {
        return class SerializableLocal extends InnerSuperClass {
        }
    };
    let ResultClass = LocalMixin(SuperClass);
    return ResultClass;
};

const AMixin = <K extends Constructor<Initable> & Initable>(SuperClass: K) => {
    let SomeHowOkay = class A extends SuperClass {
    };

    // Error!
    //
    // Type
    //
    //  {
    //      new (...args: any[]): Serializable<K>.SerializableLocal;
    //      prototype: Serializable<any>.SerializableLocal;
    //      init(...args: any[]): void; } & K
    //  } & K
    //  
    // is not a constructor function type.
    let SomeHowNotOkay = class A extends Serializable(SuperClass) {
    };
};
```

You can see that this is because `ResultClass` has the type

```
{
    new (...args: any[]): SerializableLocal;
    prototype: Serializable<any>.SerializableLocal;
    init(...args: any[]): void;
} & K
```

This is weird, because it has nothing to do with the construct signature. You'd have the same behavior if you swapped this to be `{} & K`

```diff
     let ResultClass = LocalMixin(SuperClass);
-    return ResultClass;
+    return ResultClass as {} & K;
 }
```

However, if you just leave it as a bare type parameter, TypeScript is okay. The following change will make things work!

```diff
     let ResultClass = LocalMixin(SuperClass);
-    return ResultClass;
+    return ResultClass as K;
 }
```

I guess the question is: is this intentional? Why do we disallow extending from a `{} & K` but we're okay with extending from `K`? It feels like we should probably allow this.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #33127

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot user test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -6429,11 +6431,11 @@ namespace ts {
             return true;
         }
 
-        // A valid base type is `any`, any non-generic object type or intersection of non-generic
+        // A valid base type is `any`, any non-generic object type or intersection of constructor types and non-generic
         // object types.
         function isValidBaseType(type: Type): type is BaseType {
             return !!(type.flags & (TypeFlags.Object | TypeFlags.NonPrimitive | TypeFlags.Any)) && !isGenericMappedType(type) ||
-                !!(type.flags & TypeFlags.Intersection) && every((<IntersectionType>type).types, isValidBaseType);
+                !!(type.flags & TypeFlags.Intersection) && every(filter((<IntersectionType>type).types, t => !isConstructorType(t)), isValidBaseType);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]:
```suggestion
                !!(type.flags & TypeFlags.Intersection) && every((<IntersectionType>type).types, t => isConstructorType(t) || isValidBaseType(t));
```

**Comment by weswigham** [comment]
User baselines are just today's as yet uncaptured/accepted update (just changes in upstream projects), so are clean~

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot user test this
@typescript-bot perf test this

**Pull request merged**